cni: Use Flannel instead Calico

oshoval opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Multus v4.2.1 exposed a problem of Calico on upgrade tests [1].
Since Calico is feature rich, it is also heavier and require more maintenance.
Therefore we are switching to Flannel on vm based providers.

This PR switches to Flannel and fixes the following:
1. Remove Flannel programmatic skip - Flannel didn't support Network policies, so those tests were skipped.
Now we install `kube-network-policies` pod next to Flannel to support native k8s Network policies.
2. Revert temporary WA https://github.com/kubevirt/kubevirt/pull/15349.
3. Return CNAO / Multus to sig-operator lane to address the original problem of multus + Calico that triggered this effort. 
4. Return the missing secondary interface test to the Upgrade tests.

[1] https://github.com/projectcalico/calico/issues/10723

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  

- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issue/<vep_tracking_issue_number>
-->

- Fixes #15332
- Fixes #15350

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
9. If no release note is required, just write "NONE".
-->
```release-note
None
```

oshoval commented:
[pull-kubevirt-e2e-k8s-1.33-sig-operator](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/15394/pull-kubevirt-e2e-k8s-1.33-sig-operator/1953377286732386304)

passed with flannel 0.7.0 (and skip test 1514)
soon we can test it with the latest image

oshoval commented:
/test pull-kubevirt-e2e-k8s-1.33-sig-operator

oshoval commented:
one failure here about MTU (pass locally)
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/15394/pull-kubevirt-e2e-k8s-1.33-sig-network/1954829705387446272

https://github.com/flannel-io/flannel/blob/master/Documentation/backends.md
"if not defined the MTU of the external interface is used."
we saw already issues in the past that CI has smaller MTU than locally

oshoval commented:
/test pull-kubevirt-e2e-k8s-1.33-sig-compute
/test pull-kubevirt-e2e-k8s-1.33-sig-storage
/test pull-kubevirt-e2e-k8s-1.33-sig-compute-migrations

oshoval commented:
/test pull-kubevirt-e2e-k8s-1.33-sig-compute 

not sure why KUBEVIRT_FLANNEL was false there, retrying

oshoval commented:
Hi @brianmcarey 
The 3 lanes you asked passed, and the sig-operator, sig-network as well with the fixes

https://github.com/kubevirt/kubevirtci/pull/1506 ready for review please
and then i will be able to remove the DNM commits that helped tested it here

Thanks

oshoval commented:
/test pull-kubevirt-e2e-k8s-1.33-sig-operator

oshoval commented:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/15394/pull-kubevirt-e2e-k8s-1.33-sig-operator/1955178971893075968
works (just double checked after latest updates)

ormergi commented:
Please consider conducting this transition in two phases to give use better sense whether Flannel is stable on CI:
Transition periodic jobs first, wait one week for indication its stable. If no issue occurred transition presubmit jobs.

EdDev commented on tests/operator/operator.go:
```diff
@@ -3103,6 +3104,32 @@ func generateMigratableVMIs(num int) ([]*v1.VirtualMachineInstance, error) {
+	vmi := vmis[lastVMIIndex]
+
+	const nadName = "secondarynet"
+	netAttachDef := libnet.NewBridgeNetAttachDef(nadName, nadName)
+	_, err := libnet.CreateNetAttachDef(context.Background(), testsuite.GetTestNamespace(vmi), netAttachDef)
+	Expect(err).NotTo(HaveOccurred())
```
This is an odd side-effect.
The helper itself is creating VMI objects which are later created on the K8S cluster, but this part is creating a NAD as a side effect in a hidden manner.

I suggest to do this only before actually creating the VMs, even if it means calling it several times. It is probably a fit in its own setup fixture.

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -722,13 +722,28 @@ var _ = Describe(SIG("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:
-				By("Create another VMI")
+				By("Create another VMI on same node")
 				anotherVmi = libvmifact.NewAlpineWithTestTooling(libnet.WithMasqueradeNetworking())
+				anotherVmi.Spec.Affinity = &k8sv1.Affinity{
+					PodAffinity: &k8sv1.PodAffinity{
+						RequiredDuringSchedulingIgnoredDuringExecution: []k8sv1.PodAffinityTerm{
+							{
+								LabelSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{
+										"mtu-test-primary": "true",
+									},
+								},
+								TopologyKey: "kubernetes.io/hostname",
+							},
+						},
+					},
+				}
```
Please wrap it with a builder so a (future) reader of the test can understand what it does without entering the details on how.

EdDev commented on tests/network/vmi_networking.go:
> Use pod affinity to ensure both VMs run on same node,
> this way, it avoids the cross-node tunnel,
> Allowing the test to be CNI agnostic.

Because of a specific CNI, the test coverage has been limited for several tests, which seems a bit extreme to me.

Usually an MTU validation is intended to be checked between nodes, as that provides the full coverage of a real scenario. The datagram size for the test can be calculated based on the tunnel overhead, the same way IP headers have been considered.

If these tests are not marked as "conformance", then I think we can hard-code and assume the tunnel overhead.
If one of the tests are considered "conformance", then we need to input the inter-node tunnel overhead.

With the last section in mind, I do not understand how this is working for some downstream executions. I would assume some executions on D/S are using different CNIs which also have potentially such an overhead.

oshoval replied:
It is not marked conformance (detected it only when ran on k/k itself sig-network),
kubevirtci runs conformance and passed.

We can't use hardcode value, because other CNIs (i.e calico), dont have the same phenomena
The test doesnt 2 things per IPv4, IPv6
It checks the max allowed can pass and max allowed +1 can't so it must know the exact one
only Flannel has a tunnel between the nodes that has different MTU between IPv4 and IPv6,
(and it isn't configurable per family as far as i saw)

This asymmetric makes IPv4 / IPv6 need to know both about the family and the CNI,
if we want to travel between nodes

We can do "if flannel" if desired
Please let me know how you prefer

EdDev replied:
> We can't use hardcode value, because other CNIs (i.e calico), dont have the same phenomena

We have for example OVN-K which also uses overlay/tunneling between nodes, and I know of D/S projects that run against it. So I am not clear how come they have not found this a problem until now. Unless they do not run such tests at all.

> only Flannel has a tunnel between the nodes that has different MTU between IPv4 and IPv6

So the problem is that the tunnel overhead is different between IPv4 and IPv6? This is odd.
Any idea why this is the case? Or why other overlay based network providers do not experience this?
What is the overhead of the tunnel per each stack?

BTW, IIUC you are talking about a dual-stack cluster, which for me means that the tunnel will use either IPv4 or IPv6 for the tunnel, not both, right?
Or are we talking here about single-stack vs dual-stack case?

> If we can delegate it to a follow-up it will be awesome

I am concern that we degrade the coverage with this change. MTU settings in the same node has no strong value. In reality, if two VMs communicate on the same node, they can have the maximum MTU possible as packet losses are too rare to matter.
And it also makes sense to understand where this is coming from before we workaround this one.

EdDev commented on tests/libnet/plumbing.go:
Nice catch!
I suggest you send this in its own PR, it has the potential to be taken in very quickly.

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/15460
will remove it from here

EdDev reviewed (changes requested):
Thank you, great to see this coming!

I have reviewed it although it is still in "draft".
If it is still in-progress, sorry in advance.

oshoval commented:
Thanks

Addressed the first 3 comments (as they around same commit)
and the last one (all those are marked inline)

Added a commit that allows to test it - will test it once we converge

oshoval commented:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/15394/pull-kubevirt-generate/1957086367301242880

wont pass until hold reasons are fixed

oshoval commented:
Removed the "same node" patch, set MTU to 1480 in backend solves this, testing it
more info here
https://github.com/kubevirt/kubevirtci/pull/1506/commits/42d5a127beb416b03c0facb592151fdbeafe8802

EdDev reviewed (commented):
Thank you, looks great.
LGTM

I am not explicitly adding the lgtm because it seems the last commit needs to be removed and is in place just to run the tests before kubevirtci is adding its things.

oshoval commented:
/hold cancel

kubevirtci bumped
removed the testing injection commit

@EdDev 
ptal

enp0s3 commented on tests/operator/operator.go:
```diff
@@ -981,7 +984,11 @@ var _ = Describe("[sig-operator]Operator", Serial, decorators.SigOperator, func(
 				vmis, err = generateMigratableVMIs(2)
 				Expect(err).ToNot(HaveOccurred())
-				vmis = createRunningVMIs(vmis)
+
```
The commit message states that you want to test the cluster upgrade scenario with secondary networks in VM, but this test checks kubevirt re-creation, what added value do you see in testing the NAD here?
From what I see when we delete KV all the VMs are deleted as well.

oshoval replied:
since this test uses generateMigratableVMIs, which adds the secondary interface (as was in the original test before it was [dropped](https://github.com/kubevirt/kubevirt/commit/13c81473d9ddb5114c781195b83655ae7799e436) by mistake)
it needs the NAD, and i created it just before creating the VM, due to this comment
https://github.com/kubevirt/kubevirt/pull/15394#discussion_r2280852682

are you sure that if one deletes KV it delete all the VMs ?
anyhow it is not relevant in this case, due to the above reason

follow-up might be to add the interface just to the tests requires it + create the NAD only there

enp0s3 reviewed (commented):
@oshoval Hi, looks great! I have a small q below, besides that lgtm

enp0s3 commented:
@oshoval Can you please check the CI? the [operator](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/15394/pull-kubevirt-e2e-k8s-1.33-sig-operator/1967237978082250752/artifacts/k8s-reporter/1/1_vmis.log) lane have some relevant errors on the test you've modified:
```
            "status": {
                "conditions": [
                    {
                        "type": "Ready",
                        "status": "False",
                        "lastProbeTime": "2025-09-14T15:20:02Z",
                        "lastTransitionTime": "2025-09-14T15:20:02Z",
                        "reason": "PodNotExists",
                        "message": "virt-launcher pod has not yet been scheduled"
                    },
                    {
                        "type": "Synchronized",
                        "status": "False",
                        "lastProbeTime": null,
                        "lastTransitionTime": "2025-09-14T15:20:02Z",
                        "reason": "FailedCreate",
                        "message": "failed to render launch manifest: failed to locate network attachment definition kubevirt-test-default1/secondarynet"
                    }
                ],
```

oshoval commented:
/test pull-kubevirt-manifests

oshoval commented:
Thanks @enp0s3

oshoval commented:
@EdDev ptal

oshoval commented:
Thanks

/retest
